src/threads/ThreadState.cpp

```diff
@@ -178,28 +185,41 @@ std::vector<uint8_t> Level::serialise()
     size_t thisSize = sizeof(Level);
     thisSize += nSharedVarOffsets * sizeof(uint32_t);
 
+    std::vector<uint8_t> bytes(thisSize, 0);
+
+    // Copy the level without the shared offsets (relying on the C++ class
+    // memory layout placing members at the top
     uint8_t* bytesPtr = BYTES(this);
-    std::vector<uint8_t> bytes(bytesPtr, bytesPtr + thisSize);
+    std::memcpy(bytes.data(), bytesPtr, sizeof(Level));
+
+    // Copy the shared offsets
+    if (nSharedVarOffsets > 0) {
+        std::memcpy(bytes.data() + sizeof(Level),
+                    sharedVarOffsets,
+                    nSharedVarOffsets * sizeof(uint32_t));
+    }
+
     return bytes;
 }
 
-void Level::deserialise(const Level* other)
+std::shared_ptr<Level> Level::deserialise(const std::vector<uint8_t>& bytes)
 {
-    id = other->id;
-    depth = other->depth;
-    activeLevels = other->activeLevels;
-    maxActiveLevels = other->maxActiveLevels;
-    numThreads = other->numThreads;
-    pushedThreads = other->pushedThreads;
-    wantedThreads = other->wantedThreads;
-
-    globalTidOffset = other->globalTidOffset;
-
-    nSharedVarOffsets = other->nSharedVarOffsets;
-    sharedVarOffsets = new uint32_t[nSharedVarOffsets];
-    std::memcpy(sharedVarOffsets,
-                other->sharedVarOffsets,
-                nSharedVarOffsets * sizeof(uint32_t));
```

Comment:
Relying on directly moving the underlying bytes around avoids having to explicitly set all the fields of the object.